ubi_manifest/worker/tasks/depsolver/utils.py

```diff
@@ -0,0 +1,148 @@
+import re
+from logging import getLogger
+from collections import defaultdict, deque
+from itertools import chain
+
+from pubtools.pulplib import Criteria
+from rpm import labelCompare as label_compare  # pylint: disable=no-name-in-module
+from more_executors.futures import f_return
+
+_LOG = getLogger(__name__)
+
+
+def create_or_criteria(fields, values):
+    # fields - list/tuple of fields [field1, field2]
+    # values - list of tuples [(field1 value, field2 value), ...]
+    # creates criteria for pulp query in a following way
+    # one tuple in values uses AND logic
+    # each criteria for one tuple are agregated by to or_criteria list
+    or_criteria = []
+
+    for val_tuple in values:
+        inner_and_criteria = []
+        if len(val_tuple) != len(fields):
+            raise ValueError
+        for index, field in enumerate(fields):
+
+            inner_and_criteria.append(Criteria.with_field(field, val_tuple[index]))
+
+        or_criteria.append(Criteria.and_(*inner_and_criteria))
+
+    return or_criteria
+
+
+def flatten_list_of_sets(list_of_sets):
+    out = set()
+    for one_set in list_of_sets:
+        out |= one_set
+
+    return f_return(out)
```

Comment:
This probably doesn't need to return a future, after all there is nothing async going on here.  If you made this just return 'out', and changed a couple of f_flat_map elsewhere to just f_map, that should work and be a bit simpler.